# sum up results file for Transit hmm for new bovis hiseq set

library(dplyr)

hmm_results<-read.delim("~/Data/hmm_bovis_add_19_21_genes.wig", 
                        sep="\t", header=FALSE, stringsAsFactors=F, 
                        comment.char = '#')
head(hmm_results)

results_21<-read.delim("~/Data/hmm_bovis_hiseq_21_genes.wig",
                       sep="\t", header=F, stringsAsFactors = F,
                       comment.char = "#")

compare_21<-matrix(0, nrow=nrow(hmm_results), ncol=4)
compare_21[,1]<-hmm_results$V1
compare_21[,2]<-hmm_results$V11
compare_21[,3]<-results_21$V11
compare_21[,4] <- ifelse(compare_21[,2]==compare_21[,3], "yes", "no")
View(compare_21)
## changes 395 calls



#################
# add reads together into new insertion file (add19_21.wig)

head(b19)

my_header<-c("# Generated by tpp from MbA019_R1_001.fastq, sum19 and 20 and 
variableStep chrom=Mbovis_AF2122_97
")

wig_19_21<-matrix(0,nrow(b19), 2)
wig_19_21[,1]<-b19[,1]
wig_19_21[,2]<-b19[,2] + b21[,2]
#wig_19_21[10000:10005,]
#b19[10000:10005,]
write.table(wig_19_21, "add_19_21.wig", sep="\t", quote = F, row.names = F, col.names = F)


# change to just gene and call--will affect downstream plots
bovis_hmm <- select(hmm_results, 2, 1, 11)
colnames(bovis_hmm) <- c("gene","ORF", "call")
# have to trim whitespace before gene to compare with mtb
bovis_hmm$gene<-trimws(bovis_hmm$gene)
View(bovis_hmm)
length(bovis_hmm$call)
nrow(bovis_hmm)

#this calls two for each gene, have to ask for 'unique'
## not anymore--was using faulty prot table. new one only includes cds
#length(unique(bovis_hmm$ORF))
#bovis_hmm<-unique(bovis_hmm)


essential_genes_hmm <- bovis_hmm[bovis_hmm$'call' == 'ES',2]
sort(essential_genes_hmm)
length(essential_genes_hmm)

gd_genes <- bovis_hmm[bovis_hmm$'call' == 'GD',2]
length(gd_genes)
ga_genes <- bovis_hmm[bovis_hmm$'call' == 'GA',2]
length(ga_genes)
non_ess_genes <- bovis_hmm[bovis_hmm$'call' == 'NE',2]
length(non_ess_genes)

# sum up results file for Transit hmm for new mtb hiseq set

mtb_results<-read.delim("~/Data/hmm_mtb_add_22_23_genes.wig", 
                        sep="\t", header=FALSE, stringsAsFactors=F, 
                        comment.char = '#')
head(mtb_results)


# change to just gene and call--will affect downstream plots
mtb_hmm <- select(mtb_results, 2, 1, 11)
colnames(mtb_hmm) <- c("gene","ORF", "call")

nrow(mtb_hmm)
View(mtb_hmm)
mtb_hmm[which(mtb_hmm$gene=="Rv1150"),]


#write to .csv for sharon
write.csv(mtb_hmm, "hiseq_mtb_transit.csv", quote=F)



## make table to compare genes from bovis and mtb with orthologs list
### import list of orthologous genes from malone
#orths<-read.table("~/tn_seq/data/mb_mtb_orthologs_1.txt", sep=",", stringsAsFactors = F, header=T)
#View(orths)
#nrow(orths)
#3813
## this list is missing some of the orthologs use this instead for orthologs 
malone_df<-read.delim("~/tn_seq/data/Malone_orthologs.csv", sep=",", header=FALSE, stringsAsFactors=F, skip=3)

malone_3<-select(malone_df, 2,3,4,6,11)
nrow(malone_3)
#3999
orths<-malone_3

colnames(orths)<-c("tb_ORF", "bovis_ORF", "gene", "function", "i/v")
View(orths)
head(orths)
# bovis orf / bovis call / tb orf / tb call / v or i /  lfc / p-value
hiseq_df<-matrix(0, nrow(bovis_hmm), 7)
head(hiseq_df)
for (i in 1:nrow(bovis_hmm)){
  hiseq_df[i,1]<-bovis_hmm[i,2]
  hiseq_df[i,2]<-bovis_hmm[i,3]
  if (bovis_hmm$ORF[i] %in% orths$bovis_ORF){
      hiseq_df[i,3]<-orths[which(bovis_hmm$ORF[i] == orths$bovis_ORF),1]
      hiseq_df[i,4]<-orths[which(bovis_hmm$ORF[i] == orths$bovis_ORF),3]
      if (hiseq_df[i,3] %in% mtb_hmm$ORF){
        hiseq_df[i,4]<-mtb_hmm[which(mtb_hmm$ORF==hiseq_df[i,3]),3]
      }
  }else{
      hiseq_df[i,3]<-"NA"
  }
}
bovis_hmm$ORF[1] 
mtb_hmm[3,3]
View(hiseq_df)
hiseq_df[i,3] %in% mtb_hmm$ORF
orths[which(bovis_hmm$ORF[i] == orths$bovis_ORF),1]
#[1] "Rv1588c" "Rv0095c"
#seems to be two mtb orthologs for same bovis orf? check in file (and how file was made)
# yes Mb0098c ortholog with Rv1588c and Rv0095c
#Rv1588c is orthologs with two: Mb1614c and Mb0098c

# i must have done 'unique' mbovis in txt file, so this got rid of some of the duplicates,
# but led to missing calls for some of the tb genes.

# look at tnseq_compare.R







# add in pvalue and lfc value from resampling df ()
# import results
resamp_results<-read.delim("~/Data/resampling_results/rev_resampling_output.txt", comment.char = "#", sep="\t", header=T, stringsAsFactors=F)
head(resamp_results)
colnames(resamp_results)<-c("Orf","Name",	"Desc",	"Sites", "Mean_Ctrl",
                            "Mean_Exp",	"log2FC",	"Sum_Ctrl",	"Sum_Exp", "Delta_Mean", "p_value",	"Adj_p_value")

for (i in 1:nrow(hiseq_df)){
    if (hiseq_df[i,1] %in% resamp_results$Orf){
      hiseq_df[i,7]<-resamp_results[which(resamp_results$Orf==hiseq_df[i,1]),12]
      hiseq_df[i,6]<-resamp_results[which(resamp_results$Orf==hiseq_df[i,1]),7]
    }
}

# # add star if pvalue<0.05 and lfc2>2.0
# signif_orfs<-rev_lfc_signif$Orf
# signif_orfs[1:10]
# for (i in 1:nrow(hiseq_df)){
#   if (hiseq_df[i,1] %in% signif_orfs){
#   hiseq_df[i,6]<-"*"
#   }else{
#     hiseq_df[i,6]<-""
#   }
# }
colnames(hiseq_df)<-c("Bovis ORF", "Bovis call", "TB ORF", "TB call", "ident/var", "log_fold_change", "adj pvalue")

# test to make sure calls are correct
View(hiseq_df)
nrow(hiseq_df)
#4027

# plot logfold change in volcano plot
ggplot(data=resamp_results, aes(x=log2FC, y=-log10(Adj_p_value))) + geom_point()



## make table for export
write.csv(hiseq_df, "~/Data/hiseq_ess_calls_11_03_21.csv", quote = F, row.names = F)


# make subset where pvalue<0.05




#use rev_lfc_signif orfs
s_orfs<-lfc2_signif_10$Orf
length(s_orfs)
head(hiseq_df)
lfc2_signif_table<-hiseq_df[which(hiseq_df[,1] %in% s_orfs),]
nrow(lfc2_signif_table)
#72
View(lfc2_signif_table)

write.csv(lfc2_signif_table, "lfc2_signif10.csv", quote=F, row.names = F)


## making insertion plots with ggplot

install.packages("ggplot2")
library(ggplot2)
library(dplyr)
#load up wig file insertions for bovis
b19<-read.delim("bovis_hiseq/tpp/bovis_hiseq_tpp_19.wig", header=T, comment.char = "#", sep=" ")
b20<-read.delim("bovis_hiseq/tpp/bovis_hiseq_tpp_20.wig", header=T, comment.char = "#", sep=" ")
b21<-read.delim("bovis_hiseq/tpp/bovis_hiseq_tpp_21.wig", header=T, comment.char = "#", sep=" ")

# ta sites of unique insertions (number of nonzero values in column 2)
head(b19)
nz_19<-b19[which(b19$chrom.Mbovis_AF2122_97>0),1]
length(nz_19)
#22539
nz_20<-b20[which(b20$chrom.Mbovis_AF2122_97>0),1]
length(nz_20)
#9692
nz_21<-b21[which(b21$chrom.Mbovis_AF2122_97>0),1]
length(nz_21)
#36021  (0.4898417)

total_insertion_set<-c(nz_19, nz_20, nz_21)    #not unique
length(total_insertion_set)
#68252
unique_total_insertions<-unique(total_insertion_set)
length(unique_total_insertions)
#40482/73536 = 0.551 saturation

#compare improvement with adding 19 and 21
pair_19_21_insertions<-c(nz_19, nz_21)
length(unique(pair_19_21_insertions))
#39987  (0.54 saturation)

pair_20_21_insertions<-c(nz_20, nz_21)
length(unique(pair_20_21_insertions))
#36941  (0.502 saturation)
#improves by 900 TA sites (out of total of 73536)





insertions<-b19[,2]
insertions_df<-cbind(insertions, b20[,2], b21[,2])
head(insertions_df)
totals<-rowSums(insertions_df[, c(1, 2, 3)])
head(totals)

# sum insertions into one df
positions<-b19[,1]
mb_ins_df<-data.frame(positions, totals)
#mb_ins_df<-cbind(positions, totals)
colnames(mb_ins_df)<-c("positions", "insertions")
head(mb_ins_df)
nrow(mb_ins_df)


max(mb_ins_df$insertions)
mb_ins_df[which.max(mb_ins_df[,2])]


plot(mb_ins_df$positions, mb_ins_df$insertions, type='l', main="insertion plot mbovis",
              xlab="TA position", ylab="number of insertions")

#ggplot(data = mb_ins_df) +
  #geom_bar(mapping = aes(x = positions, y = insertions), stat = "identity")

## this works better for getting pretty axes
ggplot(data=mb_ins_df, aes(x = positions, y = insertions)) +
  geom_bar(stat = "identity", width = 1, color = "black") +
  theme_bw() +
  theme(axis.text = element_text(size = 12)) +
  scale_x_continuous(name = positions, 
                     breaks = seq(0, max(positions), 500000), 
                     labels = waiver())

## make bigger plot for export
## to make bigger plot for export
png(file="mbo_insertions.png", width=2048, height=1536)
ggplot(data=mb_ins_df, aes(x = positions, y = insertions)) +
  geom_bar(stat = "identity", width = 1, color = "black") +
  theme_bw() +
  theme(axis.text.y = element_text(size = 16),
        axis.text.x = element_text(size = 14),
        axis.title=element_text(size=16,face="bold")) +
  ylim(0, 10000) +
  scale_x_continuous(name = "position", 
                     breaks = seq(0, max(positions), 500000), 
                     labels = waiver()) +
  ggtitle("Insertion plot M.bovis (hiseq)") + 
  theme(plot.title = element_text(lineheight=3, face="bold", color="black", size=20, hjust=0.5))
dev.off()


# do the same with tb
# load .wig files
b22<-read.delim("~/Data/mtb_hiseq/TPP/mtb22_hiseq_tpp.wig", header=T, comment.char = "#", sep=" ")
b23<-read.delim("~/Data/mtb_hiseq/TPP/mtb23_hiseq_tpp.wig", header=T, comment.char = "#", sep=" ")

ins_df<-data.frame(b22[,2], b23[,2])
totals<-rowSums(ins_df[, c(1,2)])
positions<-b22[,1]
mtb_ins_df<-data.frame(positions, totals)
colnames(mtb_ins_df)<-c("positions", "insertions")
head(mtb_ins_df)

ggplot(data=mtb_ins_df, aes(x = positions, y = insertions)) +
  geom_bar(stat = "identity", width = 1, color = "black") +
  theme_bw() +
  theme(axis.text = element_text(size = 12))

## to make bigger plot for export
png(file="mtb_insertions.png", width=2048, height=1536)
ggplot(data=mtb_ins_df, aes(x = positions, y = insertions)) +
  geom_bar(stat = "identity", width = 1, color = "black") +
  theme_bw() +
  theme(axis.text.y = element_text(size = 16),
        axis.text.x = element_text(size = 14),
        axis.title=element_text(size=16,face="bold")) +
  ylim(0, 10000) +
  scale_x_continuous(name = "position", 
                     breaks = seq(0, max(positions), 500000), 
                     labels = waiver()) +
  ggtitle("Insertion plot M.tb (hiseq)") + 
  theme(plot.title = element_text(lineheight=3, face="bold", color="black", size=20, hjust=0.5))
dev.off()

png('Data/resamp_plot.png')
ggplot(data=sums_resamp_results, aes(x=log2FC, y=-log10(Adj_p_value))) + geom_point()
dev.off()

png('Data/resamp_plot_dj67.png', width = 720, height = 720, res=120)
ggplot(data=dj67_resamp_results, aes(x=log2FC, y=-log10(Adj_p_value))) + geom_point()
dev.off()
